File: elixir/collab/test/collab_test.exs
defmodule CollabTest do
  use Plushie.Test.Case, app: Collab

  # -- Init -------------------------------------------------------------------

  test "name defaults to empty string" do
    assert model().name == ""
  end

  test "count defaults to 0" do
    assert model().count == 0
  end

  test "dark_mode defaults to false" do
    assert model().dark_mode == false
  end

  test "returns a Model struct" do
    assert %Collab.Model{} = model()
  end

  # -- Counter ----------------------------------------------------------------

  test "increment increases count" do
    click("#inc")
    assert model().count == 1
  end

  test "decrement decreases count" do
    click("#dec")
    assert model().count == -1
  end

  test "multiple increments accumulate" do
    click("#inc")
    click("#inc")
    click("#inc")
    assert model().count == 3
  end

  # -- Text inputs ------------------------------------------------------------

  test "name input updates name" do
    type_text("#name", "Alice")
    assert model().name == "Alice"
  end

  test "notes input updates notes" do
    type_text("#notes", "meeting at 3pm")
    assert model().notes == "meeting at 3pm"
  end

  # -- Dark mode toggle -------------------------------------------------------

  test "toggle sets dark_mode" do
    toggle("#theme")
    assert model().dark_mode == true
  end

  # -- View structure ---------------------------------------------------------

  test "has header text" do
    assert_text("#theme-root/bg/header", "Plushie Demo")
  end

  test "has name input" do
    assert_exists("#theme-root/bg/name")
  end

  test "has counter widgets" do
    assert_exists("#theme-root/bg/counter-row/inc")
    assert_exists("#theme-root/bg/counter-row/dec")
    assert_exists("#theme-root/bg/counter-row/count")
  end

  test "has theme checkbox" do
    assert_exists("#theme-root/bg/theme")
  end

  test "has notes input" do
    assert_exists("#theme-root/bg/notes")
  end

  test "initial tree matches snapshot" do
    assert :ok =
             Plushie.Test.assert_tree_snapshot(
               tree(),
               Path.join(["test", "snapshots", "collab_initial.json"])
             )
  end

  test "count text reflects model" do
    click("#inc")
    click("#inc")
    assert_text("#theme-root/bg/counter-row/count", "Count: 2")
  end

  # -- Settings ---------------------------------------------------------------

  test "has default_event_rate" do
    settings = Collab.settings()
    assert Keyword.get(settings, :default_event_rate) == 30
  end
end

File: typescript/collab/test/collab.test.ts
/**
 * Collab app tests.
 *
 * Integration tests run through the real plushie-renderer binary in
 * headless mode. Pure function tests cover init/update/view logic.
 */

import { afterAll, beforeAll, describe, expect, test } from "vitest"
import { createSession, stopPool } from "plushie/testing"
import type { TestSession } from "plushie/testing"
import collabApp, { init, update, view } from "../src/collab.js"
import type { Model } from "../src/collab.js"

// -- Integration tests (real binary) -----------------------------------------

// Tests are sequential -- shared session, no reset between tests.
describe("collab (integration)", () => {
  let session: TestSession<Model>

  beforeAll(async () => {
    session = await createSession(collabApp, { mode: "headless" })
    await session.start()
  })

  afterAll(() => {
    session?.stop()
    stopPool()
  })

  test("model starts with correct defaults", () => {
    const m = session.model()
    expect(m.name).toBe("")
    expect(m.notes).toBe("")
    expect(m.count).toBe(0)
    expect(m.darkMode).toBe(false)
    expect(m.status).toBe("")
  })

  test("window tree exists", () => {
    const tree = session.tree()
    expect(tree).not.toBeNull()
    expect(tree!.type).toBe("window")
    expect(tree!.id).toBe("main")
  })

  test("all key widgets exist", async () => {
    await session.assertExists("title")
    await session.assertExists("name")
    await session.assertExists("inc")
    await session.assertExists("dec")
    await session.assertExists("count")
    await session.assertExists("theme")
    await session.assertExists("notes")
  })

  test("status widget absent when status is empty", async () => {
    await session.assertNotExists("status")
  })

  test("clicking inc increments count", async () => {
    await session.click("inc")
    expect(session.model().count).toBe(1)

    await session.click("inc")
    expect(session.model().count).toBe(2)
  })

  test("clicking dec decrements count", async () => {
    await session.click("dec")
    expect(session.model().count).toBe(1)
  })

  test("typing in name input updates model", async () => {
    await session.typeText("name", "Alice")
    expect(session.model().name).toBe("Alice")
  })

  test("toggling dark mode checkbox", async () => {
    await session.toggle("theme")
    expect(session.model().darkMode).toBe(true)

    await session.toggle("theme")
    expect(session.model().darkMode).toBe(false)
  })

  test("title text renders", async () => {
    await session.assertText("title", "Collaborative Scratchpad")
  })
})

// -- Pure function tests (no binary) ------------------------------------------

describe("update (pure)", () => {
  test("click inc", () => expect(update(init(), "click", "inc").count).toBe(1))
  test("click dec", () => expect(update(init(), "click", "dec").count).toBe(-1))
  test("input name", () => expect(update(init(), "input", "name", "Alice").name).toBe("Alice"))
  test("toggle theme", () => expect(update(init(), "toggle", "theme").darkMode).toBe(true))
  test("unknown event unchanged", () => {
    const m = init()
    expect(update(m, "click", "nonexistent")).toBe(m)
  })
})

describe("view (pure)", () => {
  test("returns a window node", () => {
    const tree = view(init())
    expect(tree.type).toBe("window")
    expect(tree.id).toBe("main")
  })
})
